spec/range-spec.coffee

```diff
@@ -2,6 +2,139 @@ Range = require "../src/range"
 Point = require "../src/point"
 
 describe "Range", ->
+  describe "::copy()", ->
+    it "returns a copy of the given range", ->
+      expect(Range([1, 3], [3, 4]).copy().isEqual(Range([1, 3], [3, 4])))
+      expect(Range([1, 3], [2, 3]).copy().isEqual([[1, 2], [2, 3]]))
+
+  describe "::negate()", ->
+    it "should negate the start and end points", ->
+      expect(Range([ 0,  0], [ 0,  0]).negate().isEqual([[ 0,  0], [ 0,  0]]))
+      expect(Range([ 1,  2], [ 3,  4]).negate().isEqual([[-1, -2], [-3, -4]]))
+      expect(Range([-1, -2], [-3, -4]).negate().isEqual([[ 1,  2], [ 3,  4]]))
+      expect(Range([-1,  2], [ 3, -4]).negate().isEqual([[ 1, -2], [-3,  4]]))
+
+  describe "::reverse()", ->
+    it "returns a range with reversed start and end points", ->
+      expect(Range(Point(0, 2), Point(3, 4)).reverse().isEqual([[3, 4], [0, 2]]))
+      expect(Range([3, 4], [2, 3]).reverse().isEqual([[2, 3], [3, 4]]))
+
+  describe "::isEmpty()", ->
+    it "returns whether if start is equal to end", ->
+      expect(Range([0, 0], [0, 0]).isEmpty()).toBe true
+      expect(Range([2, 3], [4, 5]).isEmpty()).toBe false
+
+  describe "::isSingleLine()", ->
+    it "returns whether start row is equal to end row", ->
+      expect(Range([2, 3], [3, 4]).isSingleLine()).toBe false
+      expect(Range([1, 2], [1, 5]).isSingleLine()).toBe true
+
+  describe "::getRowCount()", ->
+    it "returns total number of rows in the given range", ->
+      expect(Range([2, 3], [4, 4]).getRowCount() is 3)
+      expect(Range([2, 4], [2, 6]).getRowCount() is 1)
+      expect(Range([2, 4], [2, 1]).getRowCount() is 1)
+
+      expect(Range([2, 5], [0, 5]).getRowCount() is -2)
+
+  describe "::getRows()", ->
+    it "returns the rows from start.row to end.row", ->
+      expect(Range([2, 5], [7, 6]).getRows() is [2..7])
+      expect(Range([2, 5], [2, 9]).getRows() is [2])
+      expect(Range([5, 6], [0, 4]).getRows() is [5..0])
+
+  describe "::freeze()", ->
+    it "makes the range object immutable", ->
+      expect(Object.isFrozen(Range([2, 4], [3, 5]).freeze()))
+      expect(Object.isFrozen(Range([0, 0], [0, 0]).freeze()))
+
+  describe "::union(otherRange)", ->
+    it "returns a new range that contains this range and the given range", ->
+      expect(Range([2, 3], [3, 3]).union(Range([3, 5], [4, 6])).isEqual([[2, 3], [4, 6]]))
+      expect(Range([2, 4], [3, 5]).union([[3, 0], [4, 5]]).isEqual([[2, 4], [4, 5]]))
+      expect(Range([2, 4], [3, 4]).union([[1, 0], [2, 7]]).isEqual([[1, 0], [3, 4]]))
+      expect(Range([2, 4], [3, 4]).union([[1, 0], [4, 5]]).isEqual([[1, 0], [4, 5]]))
+      expect(Range([2, 4], [3, 4]).union([[1, 0], [2, 2]]).isEqual([[1, 0], [3, 4]]))
+
+      expect(Range([4, 3], [2, 3]).union([[2, 2], [0, 0]]).isEqual([[2, 2], [2, 3]]))
```

Comment:
@maxbrunsfeld Due to the way `union` is implemented currently (same as text-buffer) this actually passes. I don't think it should. Should I correct the implementation?